docs/proposals/20210628-nodepool-ingress-support.md

```diff
@@ -186,44 +239,105 @@ As an end user, I want to balance the traffic into a NodePool with layer-7 load-
-	- It specifies the ingress class to NodePool name so the ingress CR can select the corresponding ingress controller.
-	- It should be granted the authorities to create/update/patch the related resources in kube-system namespace.
-	- It connects with Yurthub to monitor the resources(ingress/service/configmap/secret/endpoint) of its own NodePool.
-
-3). Ingress controller sidecar:
-	- When sidecar is launched, it gets the NodePool id it belongs to, and connects to Yurthub to list/watch the NodePool.
-	- When it detects the NodePool annotation below is changed, it will update externalIPs of the corresponding service:
-	  ```yaml
-            nodepool.openyurt.io/ingress-ips: "xxx"
-	  ```
-	- When it detects the NodePool annotation below is changed, it will update the corresponding configmap of the NodePool,
-	  which will trigger the ingress controller to reload nginx config for the resource state sync.
-	  ```yaml
-            nodepool.openyurt.io/ingress-config: "xxx"
-	  ```
+#### 1). Yurt-app-manager:
+
+Implement a standalone `YurtIngress` operator to support the NodePool ingress feature, which means that `YurtIngress`
+together with `NodePool/YurtAppSet(UnitedDeployment)/YurtAppDaemon` will compose the the current OpenYurt unitization
+capability, and they are all managed by Yurt-app-manager. The responsibility of `YurtIngress` is to monitor the update
+of `YurtIngress` CR and trigger the corresponding operations to nginx ingress controller related components.
+
+`YurtIngress` CRD definition:
+
+```go
+// YurtIngressSpec defines the desired state of YurtIngress
+type YurtIngressSpec struct {
+	// Indicates the replicas of the ingress controllers to be deployed under all the specified nodepools.
+	// +optional
+	Replicas int32 `json:"ingress_controller_replicas_per_pool,omitempty"`
+
+	// Indicates all the nodepools on which to enable ingress.
+	// +optional
+	Pools []string `json:"pools,omitempty"`
+}
+
+// YurtIngressStatus defines the observed state of YurtIngress
+type YurtIngressStatus struct {
+	// Indicates the number of the ingress controllers deployed under all the specified nodepools.
+	// +optional
+	Replicas int32 `json:"ingress_controller_replicas_per_pool,omitempty"`
+
+	// Indicates all the nodepools on which to enable ingress.
+	// +optional
+	Pools []string `json:"pools,omitempty"`
+
+	// Indicates the nginx ingress controller version deployed under all the specified nodepools.
+	// +optional
+	Version string `json:"nginx_ingress_controller_version,omitempty"`
+
+	// Total number of ready pools on which ingress is enabled.
+	// +optional
+	ReadyPoolNum int32 `json:"readyPoolNum"`
+
+	// Total number of unready pools on which ingress is enabling or enable failed.
+	// +optional
+	UnreadyPoolNum int32 `json:"unreadyPoolNum"`
+}
+```
+Some other details about the design:
+- The `YurtIngress` CR should be singleton CR which means there should be only one instance in the cluster,
```

Comment:
Only one CR can be created for the entire cluster, losing the meaning of CRD。

Reply:
If we want to operate mutli nodepools ingress simultaneously, it seems we should ensure the CR is singleton to avoid conflicts from different CRs for a same NodePool ingress. Any other suggestions will also be appreciated.